### PR TITLE
fix(html): deduplicate tags when same tag appears in both describe and test title

### LINKED ISSUE #40368: [Bug]: Tags are duplicated in HTML report
### Version

1.58.2

### Steps to reproduce

Have a describe block and a nested test both tagged with the same tag eg

```ts
test.describe("@saas tests tearsheet", () => {
  test("@saas create HTTP test using tearsheet", async ({
```

### Expected behavior

HTML reporter should only show the `@saas` tag once

### Actual behavior

Image

`saas` is here 3 times. Once because its in a `project` called saas, and then two more times. Once for the `describe` and then once for the `test`. I would expect the `describe` and `test` tags to be collapsed to one.

### Additional context

The tags in the test names already eats up a lot of screen real estate (see https://github.com/microsoft/playwright/issues/17375) and now showing multiple of the same tag pills eats up even more.

Do I have a usecase for having the same tag in the `describe()` and `test()`? Not really, this is just a huge suite with a lot of things that have moved around over the years, so I recognize this scenario may not be all that common.

### Environment

```shell
System:
    OS: macOS 26.3.1
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
    Memory: 56.99 MB / 32.00 GB
  Binaries:
    Node: 24.15.0 - /Users/seanlafferty/.nvm/versions/node/v24.15.0/bin/node
    npm: 11.12.1 - /Users/seanlafferty/.nvm/versions/node/v24.15.0/bin/npm
    pnpm: 10.33.0 - /Users/seanlafferty/.nvm/versions/node/v24.15.0/bin/pnpm
    Deno: 2.5.2 - /usr/local/bin/deno
  IDEs:
    VSCode: 1.113.0 - /usr/local/bin/code
  Languages:
    Bash: 3.2.57 - /bin/bash
  npmPackages:
    @playwright/test: 1.58.2 => 1.58.2
```

### PR DESCRIPTION
**Pull request by DoriAlagem**
Fixes #40368

When a tag like `@saas` appears in both a `test.describe()` title and the nested `test()` title, it shows up twice in the HTML reporter labels.

```ts
test.describe('@saas tests', () => {
  test('@saas create workflow', () => {});
});
// → rendered with "@saas" label appearing twice
```

**Root cause:** `TestCase.tags` extracts `/@[\S]+/g` from the full joined title path. If `@saas` is in both the describe and test title, it's picked up twice.

## Fix

Deduplicate in the HTML reporter's serialization step (`html.ts`) rather than in `TestCase.tags`, which is also consumed by grep matching.

```ts
tags: [...new Set(test.tags)],
```

## Test

Added a case inside the existing `labels` describe block in `reporter-html.spec.ts` verifying that the label count is 1 when the same tag appears in both describe and test title.

### CODE REVIEW

**Comment by DORI2001** [comment]
@microsoft-github-policy-service agree

File: packages/playwright/src/common/test.ts [hunk 1]
```diff
@@ -303,10 +303,12 @@ export class TestCase extends Base implements reporterTypes.TestCase {
     this.parent._collectTagTitlePath(path);
     path.push(this.title);
     const titleTags = path.join(' ').match(/@[\S]+/g) || [];
-    return [
+    // Deduplicate: a tag appearing in both a describe title and the test title
+    // (or via test.tag()) should only appear once in the final list.
+    return [...new Set([
```
Comment (dgozman) [inline packages/playwright/src/common/test.ts]: I don't think this is the right place to fix it. We have a strict guarantee that all tags are concatenated for grep matching, for example. The fix should be in the html reporter specifically.

File: tests/playwright-test/test-tag.spec.ts [hunk 2]
```diff
@@ -91,6 +91,42 @@ test('should have correct tags', async ({ runInlineTest }) => {
   ]);
 });
 
+test('should deduplicate tags that appear in both describe title and test title', async ({ runInlineTest }) => {
```
Comment (dgozman) [inline tests/playwright-test/test-tag.spec.ts]: Let's not add a new test, but rather amend an existing test in reporter-html that has some tags in it.

**Review by dgozman — COMMENTED** [review]
Thank you for the PR!

File: packages/playwright/src/common/test.ts (hunk 1 above)
Comment (DoriAlagem) [reply packages/playwright/src/common/test.ts]: Good point — moving the dedup to the HTML reporter makes more sense since the tags getter is also used for grep. Will update.

File: tests/playwright-test/test-tag.spec.ts (hunk 2 above)
Comment (DoriAlagem) [reply tests/playwright-test/test-tag.spec.ts]: Got it, will amend an existing one in reporter-html instead.

File: tests/playwright-test/reporter-html.spec.ts [hunk 3]
```diff
@@ -2617,6 +2617,27 @@ for (const useIntermediateMergeReport of [true, false] as const) {
         await expect(page.locator('.header-title')).toHaveText('Test passed -- @call @call-details @e2e @regression #VQ457');
         await expect(page.locator('.label')).toHaveText(['firefox', 'Monitoring', 'call', 'call-details', 'e2e', 'regression']);
       });
+
+      test('should not show duplicate labels when tag appears in both describe and test title', async ({ runInlineTest, showReport, page }) => {
```
Comment (dgozman) [inline tests/playwright-test/reporter-html.spec.ts]: Instead of adding this test, could you please just update `should show labels in the test row` in this file to have some duplicate tags?

File: tests/playwright-test/reporter-html.spec.ts [hunk 4]
```diff
@@ -2617,6 +2617,27 @@ for (const useIntermediateMergeReport of [true, false] as const) {
...
+        // Regression test for https://github.com/microsoft/playwright/issues/40368
```
Comment (dgozman) [inline tests/playwright-test/reporter-html.spec.ts]: Please add this as an annotation, there is an example in this file already:

```ts
annotation: { type: 'issue', description: 'https://github.com/microsoft/playwright/issues/40368' }
```

**Review by dgozman — COMMENTED** [review]
Thank you for the changes. The implementation looks good now, just some comments about the test.

File: tests/playwright-test/reporter-html.spec.ts (hunk 3 above)
Comment (DoriAlagem) [reply tests/playwright-test/reporter-html.spec.ts]: Done — folded the duplicate-tag scenario into the existing test by wrapping one of the tests in a `test.describe('@smoke tests', ...)` block. The existing assertion already verifies labels appear without duplicates, so it now covers this regression too.

File: tests/playwright-test/reporter-html.spec.ts (hunk 4 above)
Comment (DoriAlagem) [reply tests/playwright-test/reporter-html.spec.ts]: Added the issue annotation to the test.

**Review by dgozman — APPROVED** [review]
Looks good to me. Thank you for the PR.
